---
title: "MinION de novo assembly"
author: "Adriel Latorre-PÃ©rez"
date: "5 de junio de 2019"
output: html_document
update: "Mayo, 2020"
---

Here is all the R code needed to generate the figures used in the draft manuscript entitled "Assembly methods for nanopore-based metagenomic sequencing: an evaluation." The project consist on downloading the data generated by Nicholls et al. (2019), subsampling the datasets in order to obtain a typical MinION output (3 and 6 Gbps), and use different assemblers for evaluating the state-of-art.

All the data needed for the figures is distributed in different CSV files which could be tracked from this document.

**NOTE**: some figures has been manually modified by Inkscape in order to accomodate them to a better visualization.

# Figure 1

## A. Even Mock Community statistics for each assembler.

I want to create a Figure similar to Figure 1 of Goldstein et al. (2019), in which we can observe the performance of each assembler in a visual way.

1. Load the data.

```{r results=FALSE}
library(ggpubr)
d.fig1 = read.csv2(file = "./Fig1.csv", header = TRUE, sep = ",", dec = ".")
d.fig1$N50 = d.fig1$N50 / 1000
d.fig1
```

2. Generate the Figures.

We will try to separate each Figure in grids.

### TIME

```{r}
library(ggplot2)
p <- ggplot(d.fig1, aes(x = Time, y = Assembler, colour = Output)) + geom_point(aes(size = 2, alpha = 0.7)) + xlab("Time (s)") + ylab("")
p = p + theme_bw() + scale_y_discrete(limits = c("Canu", "metaFlye v2.4", "metaFlye v2.7", 
                             "miniasm", "Pomoxis", "Raven", 
                             "RedBean", "Shasta", "Unicycler", "Megahit", "Minia"))
p
```

### CONTIGS - LONG-READS

We separate short and long reads.

```{r}
library(ggplot2)
p2 <- ggplot(d.fig1[d.fig1$Length == "Long-reads", ], aes(x = Contigs, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("")
p2 = p2 + theme_bw() 
p2
```

### CONTIGS - SHORT-READS

```{r}
library(ggplot2)
p2.1 <- ggplot(d.fig1[d.fig1$Assembler == "Minia", ], aes(x = Contigs, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p2.1 = p2.1 + theme_bw()
p2.1
```


```{r}
library(ggplot2)
p2.2 <- ggplot(d.fig1[d.fig1$Assembler == "Megahit", ], aes(x = Contigs, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p2.2 = p2.2 + theme_bw() 
p2.2
```


```{r}
library(gridExtra)
p2.f = ggarrange(p2.2, p2.1, p2, 
          legend = FALSE, heights = c(1,1,2),
          ncol = 1, nrow = 3)
p2.f
```


### N50

```{r}
library(ggplot2)
p3 <- ggplot(d.fig1, aes(x = N50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + xlab("N50 (kbps)") + ylab("")
p3 = p3 + theme_bw() + scale_y_discrete(limits = c("Canu", "metaFlye v2.4", "metaFlye v2.7", 
                             "miniasm", "Pomoxis", "Raven", 
                             "RedBean", "Shasta", "Unicycler", "Megahit", "Minia"))
p3
```


### L50

We are going to separate long and short reads.

```{r}
library(ggplot2)
p4 <- ggplot(d.fig1[d.fig1$Length == "Long-reads", ], aes(x = L50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("")
p4 = p4 + theme_bw() 
p4
```

```{r}
library(ggplot2)
p4.1 <- ggplot(d.fig1[d.fig1$Assembler == "Minia", ], aes(x = L50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p4.1 = p4.1 + theme_bw()
p4.1
```


```{r}
library(ggplot2)
p4.2 <- ggplot(d.fig1[d.fig1$Assembler == "Megahit", ], aes(x = L50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p4.2 = p4.2 + theme_bw() 
p4.2
```


```{r}
library(gridExtra)
p4.f = ggarrange(p4.2, p4.1, p4, 
          legend = FALSE, heights = c(1,1,2),
          ncol = 1, nrow = 3)
p4.f
```

## FIGURE 1

```{r}
library(ggpubr)
fig1 = ggarrange(p, p3, p2.f, p4.f,
          labels = c("A", "B", "C"), common.legend = TRUE, heights = c(1, 1.5),
          ncol = 2, nrow = 2)
fig1
```


-----------------------------------------------------------------------------------------------------------------------

# Figure 2

1. Load the data

```{r results=FALSE}
d.fig2 = read.csv2(file = "./Fig2.csv", header = TRUE, sep = ",", dec = ".")
d.fig2$Size = d.fig2$Size / 1000000
d.fig2
```

2. Construct the figures

### Size

Size is represented as a difference between the theoretical size of all the genomes that compose the metagenome minus the real assembly size retrieved by each tool.

Maybe, I should change the theoretical size by the reference size --> PASCUAL

```{r}
library(ggplot2)
px <- ggplot(data=d.fig2, aes(x=Assembler, y=Size, fill=Output)) +
geom_bar(stat="identity", color="black", position=position_dodge()) +
  theme_bw() + ylab("Theoretical size - assembled size (Mbp)") + xlab("") + theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  
  scale_x_discrete(limits = c("Canu", "metaFlye v2.4", "metaFlye v2.7", 
                             "miniasm", "Pomoxis", "Raven", 
                             "RedBean", "Shasta", "Unicycler", "Megahit", "Minia"))
px
```

### Genome Fraction - MetaQUAST and Minimap2

```{r results=FALSE}
d.fig2 = read.csv2(file = "./Fig2_2.csv", header = TRUE, sep = ",", dec = ".")
d.fig2$Size = d.fig2$Size / 1000000
d.fig2
```

```{r}
py <- ggplot(data=d.fig2, aes(x=Output, y=Fraction, fill=Output)) +
  geom_bar(stat="identity", color="black", position=position_dodge()) +
  theme_bw() + ylab("Metagenome Assembled Fraction (%)") + xlab("") + expand_limits(y=c(0, 100)) +
  facet_grid(rows = vars(Method), cols = vars(Assembler)) + 
  geom_text(aes(label=Avg, group = Output), position = position_dodge(width = 1),vjust=-0.3,
            color="black", size=3.5)
py
```



------------------------------------------------------------------------------------------------------------------

# FIGURE 3

### Genome Fraction

```{r results=FALSE}
d.fig2.1 = read.csv2(file = "./genome_fraction.csv", header = TRUE, sep = ",", dec = ".")
d.fig2.1
```

```{r}
p <- ggplot(data=d.fig2.1, aes(x=Output, y=Fraction, fill=Genome)) +
  geom_bar(aes(alpha = 0.8), stat="identity", color="black", position=position_dodge(), show.legend = FALSE) +
  theme_bw() + ylab("Assembled Fraction (%)") + xlab("") + expand_limits(y=c(0, 105)) + 
  facet_grid(rows = vars(Assembler), cols = vars(Genome)) +
  geom_text(aes(label=Mean_Fraction), color="black", size=2.5)
p
```

---------------------------------------------------------------------------------------------------------------------

# FIGURE 5 - SNPs

### SNPs - bcftools and MuMmer

```{r results=FALSE}
d.fig5 = read.csv2(file = "./Fig5_snps.csv", header = TRUE, sep = ",", dec = ".")
d.fig5
```

bcftools

```{r}
library(ggplot2)
p <- ggplot(data=d.fig5, aes(x=Output, y=bcftools, fill=Output)) +
  geom_bar(aes(alpha=0.9), stat="identity", color="black", position=position_dodge()) +
  theme_bw() + ylab("Similarity (%)") + xlab("") + expand_limits(y=c(0, 105)) + 
  facet_grid(rows = vars(Label_bcftools), cols = vars(Assembler)) +
  geom_text(aes(label=mean_bcftools), vjust=-0.3, color="black", size=2.5)
p
```

MuMmer

```{r}
p2 <- ggplot(data=d.fig5, aes(x=Output, y=MuMmer, fill=Output)) +
  geom_bar(aes(alpha=0.9), stat="identity", color="black", position=position_dodge()) +
  theme_bw() + ylab("Similarity (%)") + xlab("") + expand_limits(y=c(0, 105)) + 
  facet_grid(rows = vars(Label_MuMmer), cols = vars(Assembler)) +
  geom_text(aes(label=mean_MuMmer), vjust=-0.3, color="black", size=2.5)
p2
```

Let's join bcftools & MuMmer figures

```{r}
library(ggpubr)
fig5.1 = ggarrange(p, p2, common.legend = TRUE, heights = c(1, 1),
          ncol = 1, nrow = 2, font.label = list(size = 11, face = "bold", color ="black", family = NULL), 
          vjust = 1, hjust = -0.2)
fig5.1
# ggsave("./Figures-April-2020/Fig5a.png", fig5.1, device = "png")
```


---------------------------------------------------------------------------------------------------------------------

# FIGURE 6 - indels

### indels - bcftools and MuMmer

```{r results=FALSE}
d.fig6 = read.csv2(file = "./Fig6_indels.csv", header = TRUE, sep = ",", dec = ".")
d.fig6
```

bcftools

```{r}
library(ggplot2)
p <- ggplot(data=d.fig6, aes(x=Output, y=bcftools, fill=Output)) +
  geom_bar(aes(alpha = 0.9), stat="identity", color="black", position=position_dodge()) +
  theme_bw() + ylab("INDELs (%)") + xlab("") + expand_limits(y=c(0, 0.65)) +
  facet_grid(rows = vars(Label_bcftools), cols = vars(Assembler)) +
  geom_text(aes(label=mean_bcftools), vjust=-0.3, color="black", size=2.5)
p
```

MuMmer

```{r}
p2 <- ggplot(data=d.fig6, aes(x=Output, y=MuMmer, fill=Output)) +
  geom_bar(aes(alpha = 0.9), stat="identity", color="black", position=position_dodge()) +
  theme_bw() + ylab("INDELs (%)") + xlab("") + expand_limits(y=c(0, 8.6)) +
  facet_grid(rows = vars(Label_MuMmer), cols = vars(Assembler)) +
  geom_text(aes(label=mean_MuMmer), vjust=-0.3, color="black", size=2.5)
p2
```

Let's join MumMer and bcftools

```{r}
library(ggpubr)
fig6.1 = ggarrange(p, p2, common.legend = TRUE, heights = c(1, 1),
          ncol = 1, nrow = 2, font.label = list(size = 11, face = "bold", color ="black", family = NULL), 
          vjust = 1, hjust = -0.2)
fig6.1
#ggsave("./Figures-April-2020/Fig5b.png", fig6.1, device = "png")
```

We save both figures (5.1 and 6.1) independently. Then we will join them by Inkscape.

