---
title: "MinION de novo assembly"
author: "Adriel Latorre-Pérez"
date: "5 de junio de 2019"
output: html_document
---

Here is all the R code needed to generate the figures used in the draft manuscript entitled "Assembly methods for nanopore-based metagenomic sequencing: an evaluation." The project consist on downloading the data generated by Nicholls et al. (2019), subsampling the datasets in order to obtain a typical MinION output (3 and 6 Gbps), and use different assemblers for evaluating the state-of-art.

All the data needed for the figures is distributed in different CSV files which could be tracked from this document.

**NOTE**: some figures has been manually modified by Inkscape in order to accomodate them to a better visualization.

**NOTE 2**: Not all figures has been used in the final manuscript due to miscellanous reasons.

# Figure 1

## A. Even Mock Community statistics for each assembler.

I want to create a Figure similar to Figure 1 of Goldstein et al. (2019), in which we can observe the performance of each assembler in a visual way.

1. Load the data.

```{r results=FALSE}
library(ggpubr)
d.fig1 = read.csv2(file = "./Fig1.csv", header = TRUE, sep = ",", dec = ".")
d.fig1$N50 = d.fig1$N50 / 1000
d.fig1
```

2. Generate the Figures.

We will try to separate each Figure in grids.

### TIME

```{r}
library(ggplot2)
p <- ggplot(d.fig1, aes(x = Time, y = Assembler, colour = Output)) + geom_point(aes(size = 2, alpha = 0.7)) + xlab("Time (s)") + ylab("")
p = p + theme_bw() + scale_y_discrete(limits = c("Canu", "metaFlye v2.4", "metaFlye v2.7", 
                             "miniasm", "Pomoxis", "Raven", 
                             "RedBean", "Shasta", "Unicycler", "Megahit", "Minia"))
p
```

### CONTIGS - LONG-READS

Como aquí tenemos grandes diferencias entre los short-reads y long-reads, me va a tocar separar los gráficos en dos partes.

```{r}
library(ggplot2)
p2 <- ggplot(d.fig1[d.fig1$Length == "Long-reads", ], aes(x = Contigs, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("")
p2 = p2 + theme_bw() 
p2
```

### CONTIGS - SHORT-READS

```{r}
library(ggplot2)
p2.1 <- ggplot(d.fig1[d.fig1$Assembler == "Minia", ], aes(x = Contigs, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p2.1 = p2.1 + theme_bw()
p2.1
```


```{r}
library(ggplot2)
p2.2 <- ggplot(d.fig1[d.fig1$Assembler == "Megahit", ], aes(x = Contigs, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p2.2 = p2.2 + theme_bw() 
p2.2
```


```{r}
library(gridExtra)
p2.f = ggarrange(p2.2, p2.1, p2, 
          legend = FALSE, heights = c(1,1,2),
          ncol = 1, nrow = 3)
p2.f
```


### N50

```{r}
library(ggplot2)
p3 <- ggplot(d.fig1, aes(x = N50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + xlab("N50 (kbps)") + ylab("")
p3 = p3 + theme_bw() + scale_y_discrete(limits = c("Canu", "metaFlye v2.4", "metaFlye v2.7", 
                             "miniasm", "Pomoxis", "Raven", 
                             "RedBean", "Shasta", "Unicycler", "Megahit", "Minia"))
p3
```


### L50

Como aquí tenemos grandes diferencias entre los short-reads y long-reads, me va a tocar separar los gráficos en dos partes.

```{r}
library(ggplot2)
p4 <- ggplot(d.fig1[d.fig1$Length == "Long-reads", ], aes(x = L50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("")
p4 = p4 + theme_bw() 
p4
```

```{r}
library(ggplot2)
p4.1 <- ggplot(d.fig1[d.fig1$Assembler == "Minia", ], aes(x = L50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p4.1 = p4.1 + theme_bw()
p4.1
```


```{r}
library(ggplot2)
p4.2 <- ggplot(d.fig1[d.fig1$Assembler == "Megahit", ], aes(x = L50, y = Assembler)) + geom_point(aes(colour = Output, size = 2, alpha = 0.7), show.legend = FALSE) + ylab("") + xlab("")
p4.2 = p4.2 + theme_bw() 
p4.2
```


```{r}
library(gridExtra)
p4.f = ggarrange(p4.2, p4.1, p4, 
          legend = FALSE, heights = c(1,1,2),
          ncol = 1, nrow = 3)
p4.f
```

## FIGURA 1

```{r}
library(ggpubr)
fig1 = ggarrange(p, p3, p2.f, p4.f,
          labels = c("A", "B", "C"), common.legend = TRUE, heights = c(1, 1.5),
          ncol = 2, nrow = 2)
fig1
# ggsave("./Log-Statistics.svg", fig1)
```


------------------------------------------------------------------------------------------------------------------

# FIGURE 3

### Genome Fraction

```{r results=FALSE}
d.fig2.1 = read.csv2(file = "./genome_fraction.csv", header = TRUE, sep = ",", dec = ".")
d.fig2.1
```

Voy a quedarme solo con las especies que tienen más de 1% en alguno de los long-reads

```{r}
d.fig2.1 = d.fig2.1[d.fig2.1$Genome == "B. subtilis" | d.fig2.1$Genome == "Li. monocytogenes" | d.fig2.1$Genome == "P. aeruginosa" | d.fig2.1$Genome == "S. cerevisiae",]
d.fig2.1
```


```{r}
library(ggplot2)
p <- ggplot(data=d.fig2.1, aes(x=Output, y=Fraction, fill=Genome)) +
  geom_bar(aes(alpha = 0.9), stat="identity", color="black", position=position_dodge(), show.legend = FALSE) +
  theme_bw() + ylab("Metagenome Assembled Fraction (%)") + xlab("") + expand_limits(y=c(0, 105)) + 
  facet_grid(rows = vars(Assembler), cols = vars(Genome)) +
  geom_text(aes(label=Mean_Fraction), vjust=-0.25, color="black", size=2.5)
p
```

